AttributeError: 'list' object has no attribute 'is_negative_prompt'

19:06:20-656180 ERROR    gradio call: AttributeError
[2023-11-21 19:06:20,656] sd: [ERROR] gradio call: AttributeError
[2023-11-21 19:06:20,656] sd: [ERROR] gradio call: AttributeError
[2023-11-21 19:06:20,656] sd: [ERROR] gradio call: AttributeError
[2023-11-21 19:06:20,656] sd: [ERROR] gradio call: AttributeError
[2023-11-21 19:06:20,656] sd: [ERROR] gradio call: AttributeError
[2023-11-21 19:06:20,656] sd: [ERROR] gradio call: AttributeError
[2023-11-21 19:06:20,656] sd: [ERROR] gradio call: AttributeError
╭────────────────────────── Traceback (most recent call last) ──────────────────────────╮
│ C:\AI\SDNEXT\automatic\modules\call_queue.py:34 in f                                  │
│                                                                                       │
│   33 │   │   │   try:                                                                 │
│ ❱ 34 │   │   │   │   res = func(*args, **kwargs)                                      │
│   35 │   │   │   │   progress.record_results(id_task, res)                            │
│                                                                                       │
│ C:\AI\SDNEXT\automatic\extensions\sd-webui-inpaint-difference\lib_inpaint_difference\ │
│ webui_nasty_hijacks.py:35 in hijack_func                                              │
│                                                                                       │
│   34 │   │                                                                            │
│ ❱ 35 │   │   return original_img2img_processing(id_task, mode, prompt, negative_promp │
│   36 │   │   │   init_img_with_mask, inpaint_color_sketch, inpaint_color_sketch_orig, │
│                                                                                       │
│ C:\AI\SDNEXT\automatic\modules\img2img.py:226 in img2img                              │
│                                                                                       │
│   225 │   │   if processed is None:                                                   │
│ ❱ 226 │   │   │   processed = processing.process_images(p)                            │
│   227 │   p.close()                                                                   │
│                                                                                       │
│ C:\AI\SDNEXT\automatic\modules\processing.py:738 in process_images                    │
│                                                                                       │
│    737 │   │   │   with context_hypertile_vae(p), context_hypertile_unet(p):          │
│ ❱  738 │   │   │   │   res = process_images_inner(p)                                  │
│    739 │   finally:                                                                   │
│                                                                                       │
│ C:\AI\SDNEXT\automatic\extensions-builtin\sd-webui-controlnet\scripts\batch_hijack.py │
│ :42 in processing_process_images_hijack                                               │
│                                                                                       │
│    41 │   │   │   # we are not in batch mode, fallback to original function           │
│ ❱  42 │   │   │   return getattr(processing, '__controlnet_original_process_images_in │
│    43                                                                                 │
│                                                                                       │
│ C:\AI\SDNEXT\automatic\modules\processing.py:859 in process_images_inner              │
│                                                                                       │
│    858 │   │   │   if shared.backend == shared.Backend.ORIGINAL:                      │
│ ❱  859 │   │   │   │   uc = get_conds_with_caching(modules.prompt_parser.get_learned_ │
│    860 │   │   │   │   c = get_conds_with_caching(modules.prompt_parser.get_multicond │
│                                                                                       │
│ C:\AI\SDNEXT\automatic\modules\processing.py:817 in get_conds_with_caching            │
│                                                                                       │
│    816 │   │   with devices.autocast():                                               │
│ ❱  817 │   │   │   cache[1] = function(shared.sd_model, required_prompts, steps)      │
│    818 │   │   cache[0] = (required_prompts, steps)                                   │
│                                                                                       │
│ C:\AI\SDNEXT\automatic\venv\lib\site-packages\sdwss\hijacker.py:12 in wrapper         │
│                                                                                       │
│   11 │   │   │   def wrapper(*args, **kwargs):                                        │
│ ❱ 12 │   │   │   │   return function(*args, **kwargs, original_function=self.__origin │
│   13                                                                                  │
│                                                                                       │
│ C:\AI\SDNEXT\automatic\venv\lib\site-packages\sdwss\sd_webui\prompt_parser.py:22 in   │
│ get_learned_conditioning                                                              │
│                                                                                       │
│    21 │   hires_steps, use_old_scheduling, *_ = args if args else (None, True)        │
│ ❱  22 │   encoder = PromptScheduleEncoder(model, prompts, steps, hires_steps, not pro │
│    23 │   conds = [encoder.process_prompt(prompt) for prompt in prompts]              │
╰───────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: 'list' object has no attribute 'is_negative_prompt'

--- a/lib_inpaint_difference/img2img_tab_extender.py
+++ b/lib_inpaint_difference/img2img_tab_extender.py
@@ -1,3 +1,4 @@
+from dataclasses import dataclass
 from gradio.context import Context as GradioContext
 from lib_inpaint_difference.ui import InpaintDifferenceTab
 

--- a/lib_inpaint_difference/img2img_tab_extender.py
+++ b/lib_inpaint_difference/img2img_tab_extender.py
@@ -1,4 +1,5 @@
 from dataclasses import dataclass
+import functools
 import gradio as gr
 from gradio.context import Context as GradioContext
 from gradio.blocks import Block as GradioBlock

--- a/lib_inpaint_difference/img2img_tab_extender.py
+++ b/lib_inpaint_difference/img2img_tab_extender.py
@@ -1,5 +1,6 @@
 from dataclasses import dataclass
 from gradio.context import Context as GradioContext
+from gradio.blocks import Block as GradioBlock
 from lib_inpaint_difference.ui import InpaintDifferenceTab
 
 

--- a/lib_inpaint_difference/img2img_tab_extender.py
+++ b/lib_inpaint_difference/img2img_tab_extender.py
@@ -1,4 +1,5 @@
 from dataclasses import dataclass
+import gradio as gr
 from gradio.context import Context as GradioContext
 from gradio.blocks import Block as GradioBlock
 from lib_inpaint_difference.ui import InpaintDifferenceTab

--- a/lib_inpaint_difference/img2img_tab_extender.py
+++ b/lib_inpaint_difference/img2img_tab_extender.py
@@ -1,0 +1,121 @@
+from gradio.context import Context as GradioContext
+from lib_inpaint_difference.ui import InpaintDifferenceTab
+
+
+class GradioContextSwitch:
+    def __init__(self, block):
+        self.block = block
+
+    def __enter__(self):
+        self.previous_block = GradioContext.block
+        GradioContext.block = self.block
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        GradioContext.block = self.previous_block
+
+
+@dataclass
+class TabData:
+    tab_index: int
+    ui_params: GradioBlock
+
+
+class Img2imgTabExtender:
+    img2img_tabs_block = None
+    ui_params_block = None
+    inpaint_params_block = None
+    amount_of_default_tabs = None
+    tab_data_list = []
+
+    @classmethod
+    def on_after_component(cls, component, **kwargs):
+        elem_id = kwargs.get('elem_id', None)
+
+        if elem_id == 'img2img_batch_inpaint_mask_dir':
+            cls.register_img2img_tabs_block(component)
+
+        if elem_id == 'img2img_mask_blur':
+            cls.register_inpaint_params_block(component)
+            cls.register_custom_ui_params_block(component)
+            cls.register_default_amount_of_tabs()
+
+            cls.create_custom_tabs()
+            cls.setup_navigation_events()
+
+    @classmethod
+    def register_img2img_tabs_block(cls, component):
+        cls.img2img_tabs_block = component.parent.parent
+
+    @classmethod
+    def register_inpaint_params_block(cls, component):
+        cls.inpaint_params_block = component.parent.parent
+
+    @classmethod
+    def register_custom_ui_params_block(cls, component):
+        cls.ui_params_block = component.parent.parent.parent
+
+    @classmethod
+    def register_default_amount_of_tabs(cls):
+        cls.amount_of_default_tabs = len(
+            [
+                child
+                for child in cls.img2img_tabs_block.children
+                if isinstance(child, gr.TabItem)
+            ]
+        )
+
+    @classmethod
+    def create_custom_tabs(cls):
+        for tab_class in [InpaintDifferenceTab]:
+            tab_index = cls._find_tab_index()
+            custom_tab_object = tab_class(tab_index)
+
+            with GradioContextSwitch(cls.img2img_tabs_block):
+                custom_tab_object.tab()
+            with GradioContextSwitch(cls.ui_params_block):
+                with gr.Group() as tab_ui_params:
+                    custom_tab_object.section()
+
+            custom_tab_object.gradio_events()
+
+            cls.register_custom_tab_data(tab_index, tab_ui_params)
+
+    @classmethod
+    def register_custom_tab_data(cls, tab_index, tab_ui_params):
+        cls.tab_data_list.append(TabData(tab_index, tab_ui_params))
+
+    @classmethod
+    def setup_navigation_events(cls):
+        img2img_tabs = [
+            child
+            for child in cls.img2img_tabs_block.children
+            if isinstance(child, gr.TabItem)
+        ]
+        for custom_tab in cls.tab_data_list:
+            for i, tab in enumerate(img2img_tabs):
+                def update_func(tab_id, custom_tab_data):
+                    return gr.update(visible=tab_id == custom_tab_data.tab_index)
+
+                tab.select(
+                    fn=functools.partial(update_func, tab_id=i, custom_tab_data=custom_tab),
+                    inputs=[],
+                    outputs=[custom_tab.ui_params]
+                )
+
+                # extend the behavior of hiding the default inpaint_params
+                if i >= cls.amount_of_default_tabs:
+                    tab.select(
+                        fn=lambda tab_id=i: gr.update(visible=False),
+                        inputs=[],
+                        outputs=[cls.inpaint_params_block]
+                    )
+
+    @classmethod
+    def _find_tab_index(cls):
+        img2img_tabs = [
+            child
+            for child in cls.img2img_tabs_block.children
+            if isinstance(child, gr.TabItem)
+        ]
+        return len(img2img_tabs)

--- a/lib_inpaint_difference/webui_callbacks.py
+++ b/lib_inpaint_difference/webui_callbacks.py
@@ -1,5 +1,6 @@
 from modules.scripts import script_callbacks
 from lib_inpaint_difference.settings import create_settings_section
+from lib_inpaint_difference.img2img_tab_extender import Img2imgTabExtender
 
 
 def on_ui_settings():
@@ -8,3 +9,4 @@ def on_ui_settings():
 
 def setup_script_callbacks():
     script_callbacks.on_ui_settings(on_ui_settings)
+    script_callbacks.on_after_component(Img2imgTabExtender.on_after_component)

--- a/scripts/inpaint_difference.py
+++ b/scripts/inpaint_difference.py
@@ -1,7 +1,6 @@
 from lib_inpaint_difference.globals import DifferenceGlobals
 from lib_inpaint_difference.webui_nasty_hijacks import hijack_img2img_processing
 from lib_inpaint_difference.webui_callbacks import setup_script_callbacks
-from lib_inpaint_difference.ui import InpaintDifferenceTab
 
 
 if DifferenceGlobals.is_extension_enabled:

--- a/scripts/inpaint_difference.py
+++ b/scripts/inpaint_difference.py
@@ -1,5 +1,3 @@
-import sdwss
-
 from lib_inpaint_difference.globals import DifferenceGlobals
 from lib_inpaint_difference.webui_nasty_hijacks import hijack_img2img_processing
 from lib_inpaint_difference.webui_callbacks import setup_script_callbacks
@@ -7,8 +5,6 @@ from lib_inpaint_difference.ui import InpaintDifferenceTab
 
 
 if DifferenceGlobals.is_extension_enabled:
-    plugin = sdwss.register_plugin("inpaint-difference")
-    plugin.append_component(InpaintDifferenceTab, name='create_img2img_tab')
     hijack_img2img_processing()
 
 

